src/main/java/im/status/keycard/KeycardApplet.java

```diff
@@ -847,6 +861,49 @@ private void loadSeed(byte[] apduBuffer) {
     resetKeyStatus();
     JCSystem.commitTransaction();
   }
+  
+  // Load up to 155 bytes of data into a buffer. This will generally be used for Shamir secrets or
+  // Bitcoin multisig scripts. The buffer (`savedData`) does not interact with anything in the applet.
+  // [PIN protected APDU]
+  private void loadData(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+    secureChannel.preprocessAPDU(apduBuffer);
+
+    if (!pin.isValidated()) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
+    }
+    
+    if (apduBuffer[ISO7816.OFFSET_LC] > SAVED_DATA_SIZE) {
+      ISOException.throwIt(ISO7816.SW_WRONG_DATA);
+    }
+    
+    JCSystem.beginTransaction();
+    Util.arrayCopy(apduBuffer, (short) ISO7816.OFFSET_CDATA, savedData, (short) 0, apduBuffer[ISO7816.OFFSET_LC]);
+    JCSystem.commitTransaction();
+    
+    secureChannel.respond(apdu, (short) 0, ISO7816.SW_NO_ERROR);
+  }
+  
+  // Export the entirety of the `savedData` buffer (155 bytes).
+  // [PIN protected APDU]
+  private void exportData(APDU apdu) {
+    byte[] apduBuffer = apdu.getBuffer();
+    secureChannel.preprocessAPDU(apduBuffer);
+
+    if (!pin.isValidated()) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
+    }
+
+    short off = SecureChannel.SC_OUT_OFFSET;
+    apduBuffer[off++] = TLV_DATA;
+    apduBuffer[off++] = (byte) SAVED_DATA_SIZE;
+    
+    JCSystem.beginTransaction();
+    Util.arrayCopy(savedData, (short) 0, apduBuffer, off++, SAVED_DATA_SIZE);
+    JCSystem.commitTransaction();
+
+    secureChannel.respond(apdu, (short) (2 + SAVED_DATA_SIZE), ISO7816.SW_NO_ERROR);
```

Comment:
This seems like an opportunity to leak secrets. If someone only saves 1 byte of data, then exports that data, they will get back their 222 bytes of the last secret that was saved. We should ensure that the data copied out exactly matches the data copied in. Either:

A) we enforce that the full 223 bytes must be written in on `loadData()`, and then the full 223 bytes will be read out with `exportData()`

or

B) We save the length of data loaded, and only export that length.

Reply:
Option A sounds safer, IMO

Reply:
Huh? That person would not need to store any data, they would just export the original. Not seeing an attack here

Reply:
Its not so much of an attack, as a data leak. Consider this:
1) user stores super secret 32 bytes in store
2) user decides not to store that data anymore, so he overwrites with an empty string (0 or 1 bytes depending on representation). As a result, the card only stores 0 or 1 byte, but the other 31 super secret bytes still exist.

From now on, any call to export data will export 1 "empty byte" followed by 31 bytes of the user's super secret data. Im not sure there is a serious attack vector here if we assume all terminals are trusted, but it definitely seems like unintentional leakage of potentially sensitive data

We really shouldn't allow `exportData()` to read out more bytes than were put in at the last call to `loadData()`

Reply:
Another potential solution is to zero-out the rest of the buffer on every call to `loadData()`. So if the user only loads 5 bytes, it will zero out the other 218 bytes for safety

Reply:
Seems like an external-application implementation detail, but it's not very important to me -- I'll switch to block smaller payloads

Reply:
998c77d